### ISSUE TITLE
What is the difference between fun and definition?

### ISSUE THREAD

**Issue by yudetamago** [LayerXcom/cbc-casper-proof#22]

**Comment by nrryuya** [comment]
Why we can't do like `definition foo :: "state * state => bool"` ?

**Comment by yudetamago** [comment]
> Why we can't do like definition foo :: "state * state => bool" ?

`definition` does not support pattern match? (ref: #10)

**Comment by yudetamago** [comment]
Where to use `definition` instead of `fun` ?
What are the merits and demerits of them?

**Comment by yudetamago** [comment]
>e.g. The conditions in the definition of is_valid_params are not used in the proof of most of the lemmas related to the safety, actually. I thought it's friendlier for Isabelle to avoid extending these definitions?

You mean `definition` is better for such a case?

**Comment by nrryuya** [comment]
I guess. 

Any thoughts, @myuon san?

**Comment by myuon** [comment]
"definition" introduces a new constant and a new axiom. For example, `definition X :: t where "X = internal_X"` is a syntax sugar of `const X` and `axiom X_def: "X = internal_X"`. If you want to prove a lemma using X and need to know the internal structure, you have to unfold it like `unfolding X_def` everytime.
On the other hand, "fun" introduces a function which compute something. Unfolding (or simplifying) the fun term is performed automatically.

In other words, `definition` is opaque for Isabelle and she doesn't know the internal structure unless you unfold it. `fun` is transparent for Isabelle. `definition` is easy to handle for prover, `fun` is easy for human. `definition` requires some effort to handle but `fun` easily messes up your goals.

NB: you can also use `primrec` for a primitive recursive function.
